casper/src/main/scala/coop/rchain/casper/rholang/InterpreterUtil.scala

```diff
@@ -47,70 +47,169 @@ object InterpreterUtil {
+                     // TODO: validate with data from config (genesis block number)
+                     maxBlockNum = 0L,
+                     // TODO: validate with sender in bonds map
+                     maxSeqNums = Map[Validator, Long](block.sender -> 0L)
+                   ).pure[F]
+                 }
+      computedParentsInfo <- computeParentsPostState(parents, preState)
+      _ <- Log[F].info(
+            s"Computed parents post state for ${PrettyPrinter.buildString(block, short = true)}."
+          )
+      result <- {
+        val (computedPreStateHash, rejectedDeployIds) = computedParentsInfo
+        if (incomingPreStateHash != computedPreStateHash) {
+          //TODO at this point we may just as well terminate the replay, there's no way it will succeed.
+          Log[F]
+            .warn(
+              s"Computed pre-state hash ${PrettyPrinter.buildString(computedPreStateHash)} does not equal block's pre-state hash ${PrettyPrinter
+                .buildString(incomingPreStateHash)}"
+            )
+            .as(none[StateHash].asRight[BlockError])
+        } else if (rejectedDeployIds.toSet != block.rejectedDeploys.toSet) {
+          // TODO: if rejected deploys are different that almost certain
+          //  hashes doesn't match also so this branch is unreachable
+          Log[F]
+            .warn(
+              s"Computed rejected deploys " +
+                s"${rejectedDeployIds.map(PrettyPrinter.buildString).mkString(",")} does not equal " +
+                s"block's rejected deploy " +
+                s"${block.rejectedDeploys.map(PrettyPrinter.buildString).mkString(",")}"
+            )
+            .as(InvalidRejectedDeploy.asLeft)
+        } else {
+          for {
+            replayResult <- replayBlock(incomingPreStateHash, block)
+            result       <- handleErrors(block.postStateHash, replayResult)
+          } yield result
+        }
+      }
+    } yield {
+      val bmd = BlockMetadata
+        .fromBlock(block)
+        .copy(
+          validated = true,
+          invalid = result.isLeft || result.right.get.isEmpty,
+          finalized = false,
+          fringe = preState.fringe.toList,
+          fringeStateHash = preState.fringeState.bytes.toArray.toByteString
+        )
+      (bmd, result)
+    }
+  }
+
+  // TODO: this is legacy code, it should be refactored with separation of errors that are
+  //  handled (with included data e.g. hash not equal) and fatal errors which should NOT be handled
   //Returns (None, checkpoints) if the block's tuplespace hash
   //does not match the computed hash based on the deploys
-  def validateBlockCheckpoint[F[_]: Concurrent: Log: BlockStore: BlockDagStorage: Span: Metrics: Timer](
-      block: BlockMessage,
-      s: CasperSnapshot,
-      runtimeManager: RuntimeManager[F]
+  def validateBlockCheckpoint[F[_]: Concurrent: Timer: RuntimeManager: BlockDagStorage: BlockStore: Log: Metrics: Span](
```

Comment:
validateBlockCheckpoint and validateBlockCheckpointNew are almost identical. The new version includes BlockMetadata in output. The old version is used is tests and should be removed when tests are fixed.